### PR TITLE
[demo] Add code for AOT tutorial

### PR DESCRIPTION
**Pull request by ailzhang**
Already passed CI in https://github.com/taichi-dev/taichi/pull/6543 (already switched to master)

We'll need to merge this PR first and then #6543 to make sure CI stability.

### CODE REVIEW

File: 0_tutorial_cgraph/app.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  ti::Runtime runtime_;
+  ti::AotModule module_;
+  ti::ComputeGraph g_demo_;
+  ti::NdArray<float> x_;
+
+  App0_tutorial() {
+    runtime_ = ti::Runtime(TI_ARCH_VULKAN);
+    module_ = runtime_.load_aot_module("0_tutorial_cgraph/assets/tutorial");
+    g_demo_ = module_.get_compute_graph("demo_graph");
+    x_ = runtime_.allocate_ndarray<float>({NPARTICLE}, {}, true);
```
Comment (jim19930609) [inline 0_tutorial_cgraph/app.cpp:18]:
```suggestion
    x_ = runtime_.allocate_ndarray<float>({NPARTICLE}, {}, true/*host_accessible*/);
```

File: 0_tutorial_cgraph/app.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    g_demo_.launch();
+    runtime_.wait();
+
+    std::vector<float> dst(NPARTICLE);
+    x_.read(dst);
+
+    float sum = std::accumulate(dst.begin(), dst.end(), 0.);
+    float expected = NPARTICLE * 50 * base;
+    if (std::abs(sum - expected) < 0.5) {
```
Comment (jim19930609) [inline 0_tutorial_cgraph/app.cpp:36]: Let's avoid magic numbers

```suggestion
    float threshold = 0.5;
    if (std::abs(sum - expected) < threshold) {
```

File: 0_tutorial_cgraph/app.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      std::cout << "Passed" << std::endl;
+    } else {
+      std::string err_str = "Expected: " + std::to_string(expected) + ", Got: " + std::to_string(sum);
+      throw std::runtime_error(err_str);
+    }
+
+    return true;
```
Comment (jim19930609) [inline 0_tutorial_cgraph/app.cpp:43]: If you throw upon failure, then it's not necessary to return a boolean

**Review by jim19930609 — APPROVED** [review]

File: 0_tutorial_cgraph/app.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  App0_tutorial() {
+    runtime_ = ti::Runtime(TI_ARCH_VULKAN);
+    module_ = runtime_.load_aot_module("0_tutorial_cgraph/assets/tutorial");
+    g_demo_ = module_.get_compute_graph("demo_graph");
+    x_ = runtime_.allocate_ndarray<float>({NPARTICLE}, {}, true);
+    std::cout << "Initialized!" << std::endl;
+  }
+
+  bool run() {
+    float base = 0.2;
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply 0_tutorial_cgraph/app.cpp:23]: Actually it's intended to use a different base/ndarray shape in C++ to demonstrate the shaders are not specialized to those inputs. 
Will add that paragraph in the tutorial instead ;)

File: 0_tutorial_cgraph/assets/tutorial.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+parser.add_argument("--arch", type=str)
+parser.add_argument("--aot", action='store_true', default=False)
+args = parser.parse_args()
+
+curr_dir = os.path.dirname(os.path.realpath(__file__))
+
+def compile_demo(arch, aot):
+    ti.init(arch=arch)
+
+    if ti.lang.impl.current_cfg().arch != arch:
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply 0_tutorial_cgraph/assets/tutorial.py:17]: It's when python taichi is not built with vulkan backend, we don't really want the automatic fallback to cpu in this case.

File: 0_tutorial_cgraph/app.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#include <cmath>
+#include <iostream>
+#include <numeric>
+#include <taichi/cpp/taichi.hpp>
+
+struct App0_tutorial {
+  static const uint32_t NPARTICLE = 8192 * 2;
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply 0_tutorial_cgraph/app.cpp:7]: same as above this is intended ;)

**Comment by ailzhang** [comment]
https://github.com/taichi-dev/taichi/pull/6551 already this passed. So I'll merge this PR first to unblock the taichi PR. ;)

**Pull request merged**